src/ext.rs

```diff
@@ -0,0 +1,51 @@
+//! HTTP extensions
+
+use bytes::Bytes;
+#[cfg(feature = "http1")]
+use http::header::{GetAll, HeaderName, IntoHeaderName};
+use http::HeaderMap;
+
+/// A map from header names to their original casing as received in an HTTP response.
+///
+/// If an HTTP/1 response `res` is parsed on a connection whose option
+/// [`http1_preserve_header_case`] was set to true and the response included
+/// the following headers:
+///
+/// ```ignore
+/// x-Bread: Baguette
+/// X-BREAD: Pain
+/// x-bread: Ficelle
+/// ```
+///
+/// Then `res.extensions().get::<HeaderCaseMap>()` will return a map with:
+///
+/// ```ignore
+/// HeaderCaseMap({
+///     "x-bread": ["x-Bread", "X-BREAD", "x-bread"],
+/// })
+/// ```
+///
+/// [`http1_preserve_header_case`]: /client/struct.Client.html#method.http1_preserve_header_case
+#[derive(Clone, Debug, Default)]
+pub struct HeaderCaseMap(HeaderMap<Bytes>);
+
+#[cfg(feature = "http1")]
+impl HeaderCaseMap {
+    /// Returns a view of all spellings associated with that header name,
+    /// in the order they were found.
+    pub fn get_all(&self, name: &HeaderName) -> GetAll<'_, Bytes> {
```

Comment:
What do you think, should we have the values be `Bytes`? Some other newtype that implies some validation has happened? Something that would allow us to change to a different representation if we wanted (`Bytes` kinda has that when we make the vtable public...)?

Reply:
I think `RawHeaderName` would be a nice addition. We can think about moving it to the `http` crate later. Or we can do it now.

Reply:
What do you propose the newtype would be able to do? Would it just be safer to start by returning some `impl Iterator<Item = impl AsRef<[u8]>>`?

Reply:
I made it return `RawHeaderName`, which only implements `Clone` and `Debug`.